parity-util-mem/src/lib.rs

```diff
@@ -85,4 +92,12 @@ mod test {
 		let s = val.malloc_size_of();
 		assert!(s > 0);
 	}
+
+	#[test]
+	fn test_dyn() {
+		trait Augmented : MallocSizeOf { }
+		impl Augmented for Vec<u8> { }
+		let val: Arc<dyn Augmented> = Arc::new(vec![0u8; 1024]);
+		assert!(malloc_size(&*val) > 1000);
```

Comment:
does it actually use `Arc` in calculation? deference implies that it doesn't?

Reply:
No, it does not, it uses those of the dereferenced trait.

This is for cases when you know that Arc is the primary reference

`MallocSizeOf` is not defined for `Arc<T: ?Sized>`

Reply:
Hmm, measuring `MallocSizeOf` for an `Arc` [is tricky](https://github.com/paritytech/parity-common/blob/f5dbb7824a93b760e447c182a6f84ec47c7a94fc/parity-util-mem/src/malloc_size.rs#L34-L44), but I wonder why a type is required to be `Sized` for `Arc` (in [`MallocUnconditionalShallowSizeOf`](https://github.com/paritytech/parity-common/blob/f5dbb7824a93b760e447c182a6f84ec47c7a94fc/parity-util-mem/src/malloc_size.rs#L229)), but not for `Box`.

Anyway, my point was that using `Arc` in tests causes a bit of confusion, we can use a trait object here directly, something like
```rust
let val = vec![0u8; 1024];
let val_ref: &dyn Augmented = &val as _; 
assert!(malloc_size(val_ref) > 1000);
```

Reply:
I am also wondering about those constraints.

About the test, I only wanted to show how to measure Arc<dyn Trait> in the test, also to ensure the exactly `Arc<dyn Trait>` can be measured -- it is a higher-level requirement, so hope we can have it as it is.